libvast/test/module.cpp

```diff
@@ -0,0 +1,688 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#define SUITE module
+
+#include "vast/module.hpp"
+
+#include "vast/aliases.hpp"
+#include "vast/concept/convertible/data.hpp"
+#include "vast/data.hpp"
+#include "vast/legacy_type.hpp"
+#include "vast/test/test.hpp"
+#include "vast/type.hpp"
+
+#include <caf/error.hpp>
+#include <caf/sum_type.hpp>
+#include <caf/test/dsl.hpp>
+#include <fmt/format.h>
+
+#include <optional>
+#include <string_view>
+
+using namespace vast;
+
+caf::expected<type>
+to_type(const std::vector<type>& known_types, const data& declaration,
+        std::string_view name = "");
+
+caf::expected<record_type>
+to_record(const std::vector<type>& known_types,
+          const std::vector<vast::data>& record_list) {
+  if (record_list.empty())
+    return caf::make_error(ec::parse_error, "parses an empty record");
+  auto record_fields = std::vector<record_type::field_view>{};
+  for (const auto& record_value : record_list) {
+    const auto* record_record_ptr = caf::get_if<record>(&record_value);
+    if (record_record_ptr == nullptr)
+      return caf::make_error(ec::parse_error, "parses an record with invalid "
+                                              "format");
+    const auto& record_record = *record_record_ptr;
+    if (record_record.size() != 1)
+      return caf::make_error(ec::parse_error, "parses an record field with an "
+                                              "invalid format");
+    auto type_or_error = to_type(known_types, record_record.begin()->second);
+    if (!type_or_error)
+      return type_or_error.error();
+    record_fields.emplace_back(record_record.begin()->first, *type_or_error);
+  }
+  return record_type{record_fields};
+}
+
+type get_known_type(const std::vector<type>& known_types,
+                    std::string_view name) {
+  for (const auto& known_type : known_types) {
+    if (name == known_type.name()) {
+      return known_type;
+    }
+  }
+  return {}; // none type
+}
+
+std::vector<std::string> reserved_names
+  = {"bool",   "integer", "count",   "real",   "duration", "time",
+     "string", "pattern", "address", "subnet", "type",     "attributes",
+     "enum",   "list",    "map",     "record", "base",     "implant",
+     "extend", "fields",  "key",     "value"};
+
+caf::expected<type> to_type(const std::vector<type>& known_types,
```

Comment:
I was thinking about (1) where we want to put this and (2) how we should name this function.

I'm thinking that maybe `type.{hpp,cpp}` make sense for the location as free functions, or as a static member of type. Regarding the naming I personally like `caf::expected<type> type::from_declaration(...)`, and would like to hear @tobim's opinion on it.